test/integration/upstreams/per_host_upstream_request.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/http/conn_pool.h"
+#include "envoy/http/metadata_interface.h"
+#include "envoy/http/protocol.h"
+#include "envoy/upstream/cluster_manager.h"
+#include "envoy/upstream/host_description.h"
+
+#include "common/http/header_map_impl.h"
+#include "common/router/upstream_request.h"
+
+#include "extensions/upstreams/http/http/upstream_request.h"
+
+namespace Envoy {
+
+namespace {
+void addHeader(Envoy::Http::RequestHeaderMap& header_map, absl::string_view header_name,
+               const envoy::config::core::v3::Metadata& metadata, absl::string_view meta_key_0,
+               absl::string_view meta_key_1) {
+  if (auto filter_metadata = metadata.filter_metadata().find(meta_key_0);
+      filter_metadata != metadata.filter_metadata().end()) {
+    const ProtobufWkt::Struct& data_struct = filter_metadata->second;
+    const auto& fields = data_struct.fields();
+    if (auto iter = fields.find(meta_key_1); iter != fields.end()) {
+      if (iter->second.kind_case() == ProtobufWkt::Value::kStringValue) {
+        header_map.setCopy(Envoy::Http::LowerCaseString(std::string(header_name)),
+                           iter->second.string_value());
+      }
+    }
+  }
+}
+} // namespace
+class PerHostHttpConnPool : public Router::GenericConnPool,
```

Comment:
Can this inherit from HttpConnPool and only override functions where different work is needed?

Reply:
Yeah, I made it